File: library-core/src/main/java/com/mikepenz/fastadapter/utils/ComparableItemListImpl.java
package com.mikepenz.fastadapter.utils;

import android.support.annotation.Nullable;

import com.mikepenz.fastadapter.IItem;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

/**
 * The default item list implementation
 */

public class ComparableItemListImpl<Item extends IItem> extends DefaultItemListImpl<Item> {

    private Comparator<Item> mComparator;

    public ComparableItemListImpl(@Nullable Comparator<Item> comparator) {
        this.mItems = new ArrayList<>();
        this.mComparator = comparator;
    }

    public ComparableItemListImpl(@Nullable Comparator<Item> comparator, List<Item> items) {
        this.mItems = items;
        this.mComparator = comparator;
    }

    /**
     * @return the defined Comparator used for this ItemAdaper
     */
    public Comparator<Item> getComparator() {
        return mComparator;
    }

    /**
     * define a comparator which will be used to sort the list "everytime" it is altered
     * NOTE this will only sort if you "set" a new list or "add" new items (not if you provide a position for the add function)
     *
     * @param comparator used to sort the list
     * @return this
     */
    public ComparableItemListImpl<Item> withComparator(@Nullable Comparator<Item> comparator) {
        return withComparator(comparator, true);
    }

    /**
     * define a comparator which will be used to sort the list "everytime" it is altered
     * NOTE this will only sort if you "set" a new list or "add" new items (not if you provide a position for the add function)
     *
     * @param comparator used to sort the list
     * @param sortNow    specifies if we use the provided comparator to sort now
     * @return this
     */
    public ComparableItemListImpl<Item> withComparator(@Nullable Comparator<Item> comparator, boolean sortNow) {
        this.mComparator = comparator;

        //we directly sort the list with the defined comparator
        if (mItems != null && mComparator != null && sortNow) {
            Collections.sort(mItems, mComparator);
            getFastAdapter().notifyAdapterDataSetChanged();
        }

        return this;
    }

    @Override
    public void move(int fromPosition, int toPosition, int preItemCount) {
        Item item = mItems.get(fromPosition - preItemCount);
        mItems.remove(fromPosition - preItemCount);
        mItems.add(toPosition - preItemCount, item);
        if (mComparator != null) {
            Collections.sort(mItems, mComparator);
        }
        getFastAdapter().notifyAdapterDataSetChanged();
    }

    @Override
    public void addAll(List<Item> items, int preItemCount) {
        mItems.addAll(items);
        if (mComparator != null) {
            Collections.sort(mItems, mComparator);
        }
        getFastAdapter().notifyAdapterDataSetChanged();
    }

    @Override
    public void addAll(int position, List<Item> items, int preItemCount) {
        mItems.addAll(position - preItemCount, items);
        if (mComparator != null) {
            Collections.sort(mItems, mComparator);
        }
        getFastAdapter().notifyAdapterDataSetChanged();
    }

    @Override
    public void setNewList(List<Item> items, boolean notify) {
        mItems = new ArrayList<>(items);
        if (mComparator != null) {
            Collections.sort(mItems, mComparator);
        }
        if(notify) {
            getFastAdapter().notifyAdapterDataSetChanged();
        }
    }
}

File: library-core/src/main/java/com/mikepenz/fastadapter/utils/ComparableItemListImpl.kt
package com.mikepenz.fastadapter.utils

import com.mikepenz.fastadapter.GenericItem
import java.util.ArrayList
import java.util.Collections
import java.util.Comparator

/**
 * The default item list implementation
 */
class ComparableItemListImpl<Item : GenericItem> @JvmOverloads constructor(
        comparator: Comparator<Item>?,
        items: MutableList<Item> = ArrayList()
) : DefaultItemListImpl<Item>(items) {

    /** @return the defined Comparator used for this ItemAdaper */
    var comparator: Comparator<Item>? = comparator
        private set

    /**
     * Define a comparator which will be used to sort the list "everytime" it is altered
     * NOTE this will only sort if you "set" a new list or "add" new items (not if you provide a position for the add function)
     *
     * @param comparator used to sort the list
     * @param sortNow    specifies if we use the provided comparator to sort now
     * @return this
     */
    @JvmOverloads
    fun withComparator(
            comparator: Comparator<Item>?,
            sortNow: Boolean = true
    ): ComparableItemListImpl<Item> {
        this.comparator = comparator

        //we directly sort the list with the defined comparator
        if (this.comparator != null && sortNow) {
            Collections.sort(_items, this.comparator)
            fastAdapter?.notifyAdapterDataSetChanged()
        }

        return this
    }

    override fun move(fromPosition: Int, toPosition: Int, preItemCount: Int) {
        val item = _items[fromPosition - preItemCount]
        _items.removeAt(fromPosition - preItemCount)
        _items.add(toPosition - preItemCount, item)
        _items.trySortWith(comparator)
        fastAdapter?.notifyAdapterDataSetChanged()
    }

    override fun addAll(items: List<Item>, preItemCount: Int) {
        _items.addAll(items)
        _items.trySortWith(comparator)
        fastAdapter?.notifyAdapterDataSetChanged()
    }

    override fun addAll(position: Int, items: List<Item>, preItemCount: Int) {
        _items.addAll(position - preItemCount, items)
        _items.trySortWith(comparator)
        fastAdapter?.notifyAdapterDataSetChanged()
    }

    override fun setNewList(items: List<Item>, notify: Boolean) {
        _items = ArrayList(items)
        _items.trySortWith(comparator)
        if (notify) {
            fastAdapter?.notifyAdapterDataSetChanged()
        }
    }
}